### PR TITLE
Support basic drag-and-drop operations

### PR DESCRIPTION
**Pull request by wineee**

### CODE REVIEW

File: src/server/qtquick/wsurfaceitem.cpp [hunk 1]
```diff
@@ -427,6 +427,12 @@ WSurface *WSurfaceItem::surface() const
 void WSurfaceItem::setSurface(WSurface *surface)
 {
     Q_D(WSurfaceItem);
+    if (!d->surface) {
+        qDebug() << "WSurfaceItem::setSurface:" << d->surface << surface;
+        releaseResources();
```
Comment (wineee) [inline src/server/qtquick/wsurfaceitem.cpp:432]: How to reset WSurfaceItem when surface become null？@zccrs

File: src/server/kernel/wcursor.cpp [hunk 2]
```diff
@@ -672,6 +672,30 @@ void WCursor::setCursorShape(CursorShape shape)
...
+    W_D(WCursor);
+    if (d->dragSurface == surface)
+        return;
+    if (d->dragSurface)
+        d->dragSurface->disconnect(this);
+
+    d->dragSurface = surface;
+    if (surface) {
+        connect(surface, &WSurface::destroyed, this, [this,d]() {
+            d->dragSurface = nullptr;
```
Comment (zccrs) [inline src/server/kernel/wcursor.cpp:686]: Needs for QPointer?

File: src/server/kernel/wcursor.cpp [hunk 3]
```diff
@@ -672,6 +672,30 @@ void WCursor::setCursorShape(CursorShape shape)
...
+            Q_EMIT dragSurfaceChanged();
+        });
+    }
+    Q_EMIT dragSurfaceChanged();
+}
+
+WSurface* WCursor::dragSurface() const
```
Comment (zccrs) [inline src/server/kernel/wcursor.cpp:693]: `WSurface *WCursor::...`

File: src/server/qtquick/wsurfaceitem.cpp [hunk 4]
```diff
@@ -427,6 +427,19 @@ WSurface *WSurfaceItem::surface() const
 void WSurfaceItem::setSurface(WSurface *surface)
 {
     Q_D(WSurfaceItem);
+
+    if (!surface) {
+        // call setSurface(nullptr) to clear resources
+        // don't check `d->surface == surface` because original surface may destroyed
+        auto tmpFlags = d->surfaceFlags;
+        d->surfaceFlags = DontCacheLastBuffer; // clear up contentItem
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp:435]: Why? If needs `DontCacheLastBuffer`, you should using it in QML.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/kernel/wcursor.cpp (hunk 2 above)
Comment (wineee) [reply src/server/kernel/wcursor.cpp:686]: not necessary

File: src/server/qtquick/wsurfaceitem.cpp (hunk 4 above)
Comment (wineee) [reply src/server/qtquick/wsurfaceitem.cpp:435]: Set in QML works for dragicon, but shouldn‘t  we clear content cache when setSurface？

File: src/server/qtquick/wsurfaceitem.cpp [hunk 5]
```diff
@@ -427,6 +427,16 @@ WSurface *WSurfaceItem::surface() const
...
+    if (!surface) {
+        // call setSurface(nullptr) to clear resources
+        // before check `d->surface == surface` because original surface maybe destroyed
+        releaseResources();
+        d->surface = nullptr;
+        Q_EMIT surfaceChanged();
+        return;
+    }
+
     if (d->surface == surface)
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp:440]: Can `if (surface && d->surface == surface)`

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Comment by wineee** [comment]
typo

**Comment by wineee** [comment]
> typo

Done

**Pull request merged**